src/persistent_merkle_tree/proof.zig

```diff
@@ -0,0 +1,113 @@
+    OutOfMemory,
+    /// Provided generalized index is not part of the binary tree (must be >= 1).
+    InvalidGindex,
+    /// Navigation toward the requested index hit a leaf before reaching the expected depth.
+    InvalidNavigation,
+    /// Witness list length does not match the gindex path length.
+    InvalidWitnessLength,
+};
+
+pub const SingleProof = struct {
+    leaf: [32]u8,
+    witnesses: [][32]u8,
+
+    pub fn deinit(self: *SingleProof, allocator: Allocator) void {
+        allocator.free(self.witnesses);
+        self.* = undefined;
+    }
+};
+
+/// Produces a single Merkle proof for the node at `gindex`.
+pub fn createSingleProof(
+    allocator: Allocator,
+    pool: *Node.Pool,
+    root: Node.Id,
+    gindex: Gindex,
+) (Node.Error || Error)!SingleProof {
+    if (@intFromEnum(gindex) < root_gindex_value) {
+        return error.InvalidGindex;
+    }
+
+    const path_len = gindex.pathLen();
+    var witnesses = try allocator.alloc([32]u8, path_len);
+    errdefer allocator.free(witnesses);
+
+    if (path_len == 0) {
+        return SingleProof{
+            .leaf = root.getRoot(pool).*,
+            .witnesses = witnesses,
+        };
+    }
+
+    var node_id = root;
+    var path = gindex.toPath();
+
+    const witness_leaf_offset: usize = 1;
+
+    for (0..path_len) |depth_idx| {
+        const witness_index = path_len - witness_leaf_offset - depth_idx;
+
+        if (path.left()) {
+            const right_id = try node_id.getRight(pool);
+            witnesses[witness_index] = right_id.getRoot(pool).*;
+            node_id = try node_id.getLeft(pool);
+        } else {
+            const left_id = try node_id.getLeft(pool);
+            witnesses[witness_index] = left_id.getRoot(pool).*;
+            node_id = try node_id.getRight(pool);
+        }
+
+        path.next();
+    }
+
+    return SingleProof{
+        .leaf = node_id.getRoot(pool).*,
+        .witnesses = witnesses,
+    };
+}
+
+/// Build a fresh node tree from a single Merkle proof.
+pub fn createNodeFromSingleProof(
+    pool: *Node.Pool,
+    gindex: Gindex,
+    leaf: [32]u8,
+    witnesses: []const [32]u8,
+) (Node.Error || Error)!Node.Id {
+    if (@intFromEnum(gindex) < root_gindex_value) {
+        return error.InvalidGindex;
+    }
+
+    const path_len = gindex.pathLen();
+    if (witnesses.len != path_len) {
+        return error.InvalidWitnessLength;
+    }
+
+    var node_id = try pool.createLeaf(&leaf, false);
+    var index_value: GindexUint = @intFromEnum(gindex);
+
+    for (witnesses) |witness| {
+        const sibling_id = try pool.createLeaf(&witness, false);
+
+        node_id = try if ((index_value & 1) == 0)
+            pool.createBranch(node_id, sibling_id, false)
+        else
+            pool.createBranch(sibling_id, node_id, false);
+        index_value >>= 1;
+    }
+
+    // Raise the reference count so callers own the result.
+    try pool.ref(node_id);
+    return node_id;
```

Comment:
I think the end result of this should be that the newly created nodes are have ref count of 1, not just the root at 1 and everything else at 0.

Reply:
My understanding is `should_ref` of `createBranch` only effects the `root`, the `left` and `right` subtree always refed by https://github.com/ChainSafe/ssz-z/blob/main/src/persistent_merkle_tree/Node.zig#L250-L251 . After tree construction, all the non root nodes ref count should already 1, only need extra call `pool.ref` for `root`.

Reply:
Ah right 👍